src/combinatorics/young_diagram.lean

```diff
@@ -123,4 +126,122 @@ end distrib_lattice
+
+@[simp] lemma transpose_eq_iff {μ ν : young_diagram} : μ.transpose = ν ↔ μ = ν.transpose :=
+by { split; { rintro rfl, ext, simp } }
+
+@[simp] lemma transpose_transpose {μ : young_diagram} : μ.transpose.transpose = μ :=
+by rw transpose_eq_iff
+
+-- This is effectively both directions of the iff statement below.
+protected lemma le_of_transpose_le {μ ν : young_diagram} (h_le : μ.transpose ≤ ν) :
+  μ ≤ ν.transpose :=
+λ c hc, by { simp only [mem_transpose], apply h_le, simpa }
+
+@[simp] lemma transpose_le_iff {μ ν : young_diagram} : μ.transpose ≤ ν ↔ μ ≤ ν.transpose :=
+⟨ young_diagram.le_of_transpose_le,
+  by { convert @young_diagram.le_of_transpose_le μ.transpose ν.transpose, simp } ⟩
+
+/-- Transposing Young diagrams is an `order_iso`. -/
+def transpose_order_iso : young_diagram ≃o young_diagram :=
+⟨⟨transpose, transpose, λ _, by simp, λ _, by simp⟩, by simp⟩
+
+end transpose
+
+section rows
+/-- Rows and row lengths of Young diagrams.
+
+This section defines `μ.row` and `μ.row_len`, with the following API:
+      1.  `(i, j) ∈ μ ↔ j < μ.row_len i`
+      2.  `μ.row i = {i} ×ˢ (finset.range (μ.row_len i))`
+      3.  `μ.row_len i = (μ.row i).card`
+      4.  `∀ {i1 i2}, i1 ≤ i2 → μ.row_len i2 ≤ μ.row_len i1`
+
+Note: #3 is not convenient for defining `μ.row_len`; instead, `μ.row_len` is defined
+as the smallest `j` such that `(i, j) ∉ μ`. --/
+
+def row (μ : young_diagram) (i : ℕ) := μ.cells.filter (λ c, c.fst = i)
+
+lemma mem_row_iff {μ : young_diagram} {i : ℕ} {c : ℕ × ℕ} : c ∈ μ.row i ↔ c ∈ μ ∧ c.fst = i :=
+by simp [row]
+
+protected lemma exists_not_mem_row (μ : young_diagram) (i : ℕ) : ∃ j, (i, j) ∉ μ.cells :=
+begin
+  obtain ⟨j, hj⟩ := infinite.exists_not_mem_finset
+    ((μ.cells).preimage (prod.mk i) (λ _ _ _ _ h, by {cases h, refl})),
+  rw finset.mem_preimage at hj, exact ⟨j, hj⟩,
+end
+
+/-- Length of a row of a Young diagram --/
+def row_len (μ : young_diagram) (i : ℕ) : ℕ := nat.find $ μ.exists_not_mem_row i
+
+lemma mem_iff_lt_row_len {μ : young_diagram} {i j : ℕ} : (i, j) ∈ μ ↔ j < μ.row_len i :=
+by { rw [row_len, nat.lt_find_iff], push_neg,
+     exact ⟨λ h _ hmj, μ.up_left_mem (by refl) hmj h, λ h, h _ (by refl)⟩ }
+
+lemma row_eq_prod {μ : young_diagram} {i : ℕ} : μ.row i = {i} ×ˢ (finset.range (μ.row_len i)) :=
+by { ext ⟨a, b⟩,
+     simp only [finset.mem_product, finset.mem_singleton, finset.mem_range,
+                mem_row_iff, mem_iff_lt_row_len, and_comm, and.congr_right_iff],
+     rintro rfl, refl }
+
+lemma row_len_eq_card (μ : young_diagram) {i : ℕ} : μ.row_len i = (μ.row i).card :=
+by simp [row_eq_prod]
+
+lemma row_len_decr (μ : young_diagram) (i1 i2 : ℕ) (hi : i1 ≤ i2) : μ.row_len i2 ≤ μ.row_len i1 :=
+by { by_contra' h_lt, rw ← lt_self_iff_false (μ.row_len i1),
+     rw ← mem_iff_lt_row_len at h_lt ⊢,
+     exact μ.up_left_mem hi (by refl) h_lt }
+
+end rows
+
+section columns
+/-- Columns and column lengths of Young diagrams.
+
+This section has an identical API to the rows section. --/
```

Comment:
Is there a way to automatically deduce each proof in this section by replacing `μ` by `μ.transpose`?

Currently this section is almost identical to the rows section.

Reply:
Maybe you could write enough lemmas to carry the theory of `col` to `transpose.row` then use `simp`? One feature you might use are [simp sets](https://leanprover-community.github.io/mathlib_docs/commands.html#mk_simp_attribute), so then you can for example write `simp with young_diagram_col_to_row` to invoke these lemmas.

Reply:
I have a draft, but this is a big enough change that I'm not sure I want to commit it directly. How should I proceed here?

Reply:
@kmill Any answer on this?